src/Foundatio.Redis/Cache/RedisCacheClient.cs

```diff
@@ -80,52 +84,50 @@ public async Task<int> RemoveAllAsync(IEnumerable<string> keys = null)
 
                 try
                 {
+                    long dbSize = await server.DatabaseSizeAsync().AnyContext();
                     await server.FlushDatabaseAsync().AnyContext();
+                    deleted += dbSize;
                     continue;
                 }
                 catch (Exception ex)
                 {
-                    if (_logger.IsEnabled(LogLevel.Error))
-                        _logger.LogError(ex, "Unable to flush database");
+                    _logger.LogError(ex, "Unable to flush database: {Message}", ex.Message);
                 }
 
                 try
                 {
+                    // NOTE: We need to use a HashSet to avoid duplicate counts due to SCAN is non-deterministic.
+                    // A Performance win could be had if we are sure dbSize didn't fail and we know nothing was changing
+                    // keys while we were deleting.
+                    var seen = new HashSet<RedisKey>();
                     await foreach (var key in server.KeysAsync().ConfigureAwait(false))
-                        await Database.KeyDeleteAsync(key).AnyContext();
+                        seen.Add(key);
+
+                    foreach (var batch in seen.Batch(batchSize))
+                        deleted += await Database.KeyDeleteAsync(batch.ToArray()).AnyContext();
```

Comment:
We could do this to basically get very low memory footprint but then the key count isn't deterministic as scan can return duplicates.

```
await foreach (var redisKeys in server.KeysAsync().BatchAsync(batchSize).ConfigureAwait(false))
   deleted += await Database.KeyDeleteAsync(redisKeys).AnyContext();
```